### ISSUE TITLE
[BUG] Debugging the Cloud Run node hello-world app doesnt work

### ISSUE THREAD

**Issue by etanshaul** [GoogleCloudPlatform/cloud-code-samples#261]
1) run the [CR node hello-world application](https://github.com/GoogleCloudPlatform/cloud-code-samples/tree/cloud-run-updates/nodejs/nodejs-cloud-run-hello-world) using `skaffold debug`

expected: debugging works
actual: application starts up with errors:

```
Port forwarding pod/node-local-t-5f558f75b-hbmgz in namespace default, remote port 8080 -> address 127.0.0.1 port 8081
Port forwarding pod/node-local-t-5f558f75b-hbmgz in namespace default, remote port 9229 -> address 127.0.0.1 port 9229
[node-local-t-5f558f75b-hbmgz node-local-t-container] Debugger listening on ws://127.0.0.1:9229/2cae197c-b3fa-4e05-aeeb-244df89b3205
[node-local-t-5f558f75b-hbmgz node-local-t-container] For help, see: https://nodejs.org/en/docs/inspector
[node-local-t-5f558f75b-hbmgz node-local-t-container] 
[node-local-t-5f558f75b-hbmgz node-local-t-container] > nodejs-cloud-run-hello-world@0.0.1 debug /usr/src/app
[node-local-t-5f558f75b-hbmgz node-local-t-container] > node --inspect=9229 index.js
[node-local-t-5f558f75b-hbmgz node-local-t-container] 
[node-local-t-5f558f75b-hbmgz node-local-t-container] Starting inspector on 127.0.0.1:9229 failed: address already in use
[node-local-t-5f558f75b-hbmgz node-local-t-container] npm ERR! code ELIFECYCLE
[node-local-t-5f558f75b-hbmgz node-local-t-container] npm ERR! errno 12
[node-local-t-5f558f75b-hbmgz node-local-t-container] npm ERR! nodejs-cloud-run-hello-world@0.0.1 debug: `node --inspect=9229 index.js`
[node-local-t-5f558f75b-hbmgz node-local-t-container] npm ERR! Exit status 12
[node-local-t-5f558f75b-hbmgz node-local-t-container] npm ERR! 
[node-local-t-5f558f75b-hbmgz node-local-t-container] npm ERR! Failed at the nodejs-cloud-run-hello-world@0.0.1 debug script.
[node-local-t-5f558f75b-hbmgz node-local-t-container] npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[node-local-t-5f558f75b-hbmgz node-local-t-container] 
[node-local-t-5f558f75b-hbmgz node-local-t-container] npm ERR! A complete log of this run can be found in:
[node-local-t-5f558f75b-hbmgz node-local-t-container] npm ERR!     /root/.npm/_logs/2020-03-31T22_50_34_271Z-debug.log
Forwarding container node-local-t-5f558f75b-hbmgz/node-local-t-container to local port 8081.
```

(note: running with `skaffold dev` works fine)

Replacing, in the [Dockerfile](https://github.com/GoogleCloudPlatform/cloud-code-samples/blob/cloud-run-updates/nodejs/nodejs-cloud-run-hello-world/Dockerfile):
`CMD ["npm", "run", "debug"]`

With (which is how our Kubernetes samples are configured):
`ENTRYPOINT ["node", "--inspect=9229", "index.js"]`

Fixes the issue

**Comment by etanshaul** [comment]
@balopat to see if this is an issue on Skaffold's side as well.

note: in order to run this app with Skaffold, you need to create k8s manifests. My testing was via the IDE's integration which generates this automatically for you. For testing, you can use (generated via `gcloud code export`):

```
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    service: node-local-t
  name: node-local-t
spec:
  replicas: 1
  selector:
    matchLabels:
      app: node-local-t
  template:
    metadata:
      labels:
        app: node-local-t
    spec:
      containers:
      - env:
        - name: PORT
          value: '8080'
        - name: K_CONFIGURATION
          value: node-local-t
        - name: K_REVISION
          value: local
        - name: K_SERVICE
          value: node-local-t
        image: node-local-t
        name: node-local-t-container
        ports:
        - containerPort: 8080
        resources:
          limits:
            cpu: 1.0
            memory: 268435456
      terminationGracePeriodSeconds: 0
---
apiVersion: v1
kind: Service
metadata:
  name: node-local-t
spec:
  ports:
  - port: 8080
    protocol: TCP
    targetPort: 8080
  selector:
    app: node-local-t
  type: LoadBalancer
```

**Comment by briandealwis** [comment]
Oh… this is because the `node:12.16.0-alpine3.10` image has an entrypoint that invokes a `docker-entrypoint.sh` launcher script.  This entrypoint script lets someone run `docker run <node-derived-image> index.js` and have it run `node index.js`.

`skaffold debug` doesn't know about this `docker-endpoint.sh`, and so it treats this image as having a script that launches `node` somehow.  So it transforms the container environment to add:

      - name: NODE_OPTIONS
          value: --inspect=9229
  
which is picked up by `node`.

But this example causes the script to run `npm run debug`, and so the `NODE_OPTIONS` is picked up by `npm`.

It's unlikely the `node` image will move away from using this `docker-entrypoint.sh`, so `skaffold debug` should be taught to handle `docker-endpoint.sh`. I've filed GoogleContainerTools/skaffold#3903 to track this.

But this image should invoke `node` directly, IMHO.  Using `npm` just adds unnecessary latency and memory pressure on startup.  This would allow `skaffold debug` to do the right thing.

**Comment by etanshaul** [comment]
thanks @briandealwis for the thorough investigation. 

@averikitsch @grayside , could you consider updating the entrypoint to invoke `node` instead of `npm` (fwiw this would also align with our k8s samples).

**Comment by grayside** [comment]
I'm not sure how the current implementation is blocking the propagation of environment variables. There's no guarantee that a node container will launch a node service in the ENTRYPOINT, in fact the general best practice is to use CMD for this purpose. ENTRYPOINT might be used for runtime configuration templating or other setup steps.

Is the problem that `npm run debug` is itself setting `--inspect=9229` and `skaffold debug` doesn't anticipate that?

The existing use of npm scripts is aligned with broad community practices around using package.json as a lightweight task management system. I'd prefer not to change that but can do so. Switching to overriding the ENTRYPOINT seems like it's tailoring the sample to an overly brittle interpretation of how developers are building containers.

**Comment by briandealwis** [comment]
`npm` is itself a `node` application and so it intercepts the `NODE_OPTIONS` environment settings, rather than the actual application.

[Google's best practices for building containers](https://cloud.google.com/solutions/best-practices-for-building-containers) recommends reducing a container's attack surface by avoiding unnecessary tools and treating containers as immutable.  I cannot see any compelling reason why  we would advocate that an application should be launched from a _build_ tool that is not required at runtime, especially when the actual command line is as simple as that needed to run the build tool (`node index.js` vs `npm start`).

**Comment by grayside** [comment]
Ah, npm running with this NODE_OPTION clarifies, thank you.

**Comment by simonz130** [comment]
This is fixed by #267
